test/compiler/immutablearray.jl

```diff
@@ -0,0 +1,499 @@
+import Core: ImmutableArray
+import Core.Compiler: arrayfreeze_tfunc, mutating_arrayfreeze_tfunc, arraythaw_tfunc
+const ImmutableVector{T} = Core.ImmutableArray{T,1}
+
+@testset "ImmutableArray tfuncs" begin
+    @test arrayfreeze_tfunc(Vector{Int}) === ImmutableVector{Int}
+    @test arrayfreeze_tfunc(Vector) === ImmutableVector
+    @test arrayfreeze_tfunc(Array) === ImmutableArray
+    @test arrayfreeze_tfunc(Any) === ImmutableArray
+    @test arrayfreeze_tfunc(ImmutableVector{Int}) === Union{}
+    @test arrayfreeze_tfunc(ImmutableVector) === Union{}
+    @test arrayfreeze_tfunc(ImmutableArray) === Union{}
+    @test mutating_arrayfreeze_tfunc(Vector{Int}) === ImmutableVector{Int}
+    @test mutating_arrayfreeze_tfunc(Vector) === ImmutableVector
+    @test mutating_arrayfreeze_tfunc(Array) === ImmutableArray
+    @test mutating_arrayfreeze_tfunc(Any) === ImmutableArray
+    @test mutating_arrayfreeze_tfunc(ImmutableVector{Int}) === Union{}
+    @test mutating_arrayfreeze_tfunc(ImmutableVector) === Union{}
+    @test mutating_arrayfreeze_tfunc(ImmutableArray) === Union{}
+    @test arraythaw_tfunc(ImmutableVector{Int}) === Vector{Int}
+    @test arraythaw_tfunc(ImmutableVector) === Vector
+    @test arraythaw_tfunc(ImmutableArray) === Array
+    @test arraythaw_tfunc(Any) === Array
+    @test arraythaw_tfunc(Vector{Int}) === Union{}
+    @test arraythaw_tfunc(Vector) === Union{}
+    @test arraythaw_tfunc(Array) === Union{}
+end
+
+@testset "ImmutableArray allocation optimization" begin
+    @noinline function op(a::AbstractArray)
+        return reverse(reverse(a))
+    end
+
+    function allo1()
+        a = Vector{Float64}(undef, 5)
+        for i = 1:5
+            a[i] = i
+        end
+        return Core.ImmutableArray(a)
+    end
+
+    function allo2()
+        a = [1,2,3,4,5]
+        return Core.ImmutableArray(a)
+    end
+
+    function allo3()
+        a = Matrix{Float64}(undef, 5, 2)
+        for i = 1:5
+            for j = 1:2
+                a[i, j] = i + j
+            end
+        end
+        return Core.ImmutableArray(a)
+    end
+
+    function allo4() # sanity check
+        return Core.ImmutableArray{Float64}(undef, 5)
+    end
+
+    function allo5() # test that throwing boundserror doesn't escape
+        a = [1,2,3]
+        try
+            getindex(a, 4)
+        catch end
+        return Core.ImmutableArray(a)
+    end
+
+    function allo6()
+        a = ones(5)
+        a = op(a)
+        return Core.ImmutableArray(a)
+    end
+
+    function test_allo()
+        # warmup
+        allo1(); allo2(); allo3();
+        allo4(); allo5(); allo6();
+
+        # these magic values are what the mutable array version would allocate
+        @test @allocated(allo1()) == 96
+        @test @allocated(allo2()) == 96
+        @test @allocated(allo3()) == 144
+        @test @allocated(allo4()) == 96
+        @test @allocated(allo5()) == 160
+        @test @allocated(allo6()) == 288
+    end
+
+    test_allo()
+end
```

Comment:
I think we probably want to refactor our test cases. Here are few things I would like to fix:
- test using `@allocated` is a bit too high-level; we should also have some lower level tests that inspects IRs (which are more reliable way to detect regression
- test using hard-coded numbers of allocated memory is not very reliable way to test: we should first obtain the number when the optimization doesn't work, and compare our test case with that number

Reply:
8507643 looks good, thanks for looking at this. I had been trying to find an elegant way to test that the allocations were the same between the mutable and immutable paths while also checking that the actual optimization is happening correctly, and this does just that :pray: